src/ILVerify/src/ILImporter.Verify.cs

```diff
@@ -1233,6 +1239,22 @@ void StartImportingInstruction()
 
         void EndImportingInstruction()
         {
+            var currentOpcode = (ILOpcode)_ilBytes[_currentInstructionOffset];
+
+            // tail.call may only be followed by return
+            if (_pendingTailcall)
+            {
+                Check(currentOpcode == ILOpcode.ret, VerifierError.TailRet);
```

Comment:
The `ret` can be in a different basic block. The ECMA spec says: "Correct CIL shall not branch to the call instruction, but it is permitted to branch to the ret".

I think that the tailcall pattern needs to be verified a part of the `call/calli/callvirt`.

(Or the pendingTailCall flag needs to be part of the basicblock entry state - but I do not think it would be a good design choice.)

Reply:
True, the test case I have added for this scenario is too simple. My implementation could report an error depending on the order the basic blocks are verified.

I will add a test case with a more complex flow graph and change this to be checked inside `ImportCall`, that's actually a lot more elegant anyway.